### PR TITLE
[Auto Parallel] Add the support for the auto completion of while_op

### PR DESCRIPTION
**Pull request by aoyulong**
### PR types

New features 

### PR changes

Others

### Describe

-  This pr can automatically complete the distributed attributes for while_op and its sub_block.
- The process_mesh completion algorithm is also improved to give users more flexibility to annotate the process_mesh.
- Fix bugs in the compatible check of dist_default op.

### CODE REVIEW

**Review by phlrain — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import paddle
+import numpy as np
+import paddle.nn as nn
+import paddle.utils as utils
+import paddle.fluid as fluid
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py]: 外部开发者不推荐使用fluid下的api，建议单测跟外部使用方式一致

File: python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        mlp_start = MLPLayer(
+            hidden_size=hidden_size,
+            intermediate_size=4 * hidden_size,
+            dropout_ratio=0.1,
+            initializer_range=0.02)
+        pred = mlp_start(input)
+
+        input_array = fluid.layers.array_write(pred, i)
+        cond = fluid.layers.less_than(x=i, y=loop_len)
+        while_op = fluid.layers.While(cond=cond)
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py]: 这里是否可以使用paddle.static.nn.while_loop？
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/static/nn/while_loop_cn.html#while-loop

File: python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py (hunk 1 above)
Comment (aoyulong) [reply python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py]: Done.

File: python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py (hunk 2 above)
Comment (aoyulong) [reply python/paddle/fluid/tests/unittests/auto_parallel/test_while_op_completion.py]: Done.

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**